src/races/races.service.ts

```diff
@@ -294,4 +296,534 @@ export class RacesService {
+        city = city.replace(/ City$/, '')
+        // Note: we don't remove Town/Township/Village/Borough
+        // because we want to keep that info for ai column matching.
+      }
+      if (county && county !== '') {
+        county = county.replace(/ County$/, '')
+      }
+
+      if (
+        (electionLevel === 'city' && !city) ||
+        (electionLevel === 'county' && !county)
+      ) {
+        logger.log(
+          slug,
+          'could not find location from mtfcc. getting location from AI',
+        )
+
+        // If we couldn't get city/county with mtfcc/geo then use the AI.
+        locationResp = await this.extractLocationAi(
+          officeName + ' - ' + electionState,
+          level,
+        )
+        logger.log(slug, 'locationResp', locationResp)
+      }
+
+      if (locationResp?.level) {
+        if (locationResp.level === 'county') {
+          county = locationResp.county
+        } else {
+          if (
+            locationResp.county &&
+            locationResp.hasOwnProperty(locationResp.level)
+          ) {
+            city = locationResp[locationResp.level]
+            county = locationResp.county
+          }
+        }
+      }
+    }
+
+    if (county) {
+      logger.log(slug, 'Found county', county)
+    }
+    if (city) {
+      logger.log(slug, 'Found city', city)
+    }
+
+    let priorElectionDates: string[] = []
+    if (partisanType !== 'partisan' && findElectionDates) {
+      // todo: restore this code once we have Position / BallotPosition
+      // if (positionId) {
+      //   const ballotPosition = await BallotPosition.findOne({
+      //     ballotHashId: positionId.toString(),
+      //   })
+      //   if (ballotPosition && ballotPosition?.ballotId) {
+      //     const ballotPositionId = ballotPosition?.ballotId
+      //     priorElectionDates = await getElectionDatesPosition(
+      //       slug,
+      //       ballotPositionId,
+      //     )
+      //     logger.log(
+      //       `priorElectionDates from PositionId ${ballotPositionId} `,
+      //       priorElectionDates,
+      //     )
+      //   }
+      // }
+
+      if ((!priorElectionDates || priorElectionDates.length === 0) && zip) {
+        priorElectionDates = await getElectionDates(
+          slug,
+          officeName,
+          zip,
+          race?.position?.level,
+        )
+        logger.log(`priorElectionDates from zip ${zip}`, priorElectionDates)
+      }
+    }
+
+    data.slug = slug
+    data.officeName = officeName
+    data.electionDate = electionDate
+    data.electionTerm = termLength
+    data.electionLevel = electionLevel
+    data.electionState = electionState
+    data.electionCounty = county
+    data.electionMunicipality = city
+    data.subAreaName = subAreaName
+    data.subAreaValue = subAreaValue
+    data.partisanType = partisanType
+    data.priorElectionDates = priorElectionDates
+    data.positionId = positionId
+    data.tier = tier
+    return data
```

Comment:
Instead of having this in one massive method, we _could_ split these out into separate `private` service methods, or even module-level util methods, depending on complexity and dependency requirements.

If we don't want to do that right now, can we at least add a `// TODO:` here to do just that?